### PR TITLE
Fixes pharo-project#3430: ifError should accept error instance instead of its properties

### PR DESCRIPTION
**Pull request by kamk**
ifError: block accepts anError instance as a first argument

### CODE REVIEW

File: src/Kernel/BlockClosure.class.st
Comment (dionisiydk) [inline src/Kernel/BlockClosure.class.st:345]: I would also remove second parameter. It is incompatible change anyway. So it should not add more problems.
(my idea was to make ifError consistent to on:do:)

**Review by dionisiydk — APPROVED** [review]

File: src/Kernel/BlockClosure.class.st
Comment (kamk) [reply src/Kernel/BlockClosure.class.st:345]: Attempt in the commit kamk/pharo@415dfdefea0d5da0a49f7da62dff382687248ef7.

**Review by dionisiydk — APPROVED** [review]

**Comment by dionisiydk** [comment]
benchmarks step is completely broken. Can we disable it? (@MarcusDenker ?)

**Comment by tesonep** [comment]
@dionisiydk I have disabled

**Comment by dionisiydk** [comment]
Great, thanks

пн, 13 янв. 2020 г., 16:23 Pablo Tesone <notifications@github.com>:

**Comment by dionisiydk** [comment]
The failure looks unrelated. CI needs to be run again

**Review by MarcusDenker — APPROVED** [review]
I like this change. But as it is not backward compatible, I will wait for more comments before merging

**Comment by Bajger** [comment]
Hi! There are about 33 senders of #ifError: message in the system. Do you expect that this message is widely used in other user's code? Could we use deprecation mechanism, when #ifError: is called with old-way arguments? (Method name #ifError: is same, but argument block would need to be tested appropriately and deprecation called if block argument count is more than 1,and if first (and only) parameter is error instance?)

**Comment by kamk** [comment]
I guess it would require some handler's code analysis if first argument is treated as some kind of String. I have added deprecation notice if more than one argument is passed.

**Comment by guillep** [comment]
> I guess it would require some handler's code analysis if first argument is treated as some kind of String. I have added deprecation notice if more than one argument is passed.

The more """organized""" way to do this would be:
 - add a new `ifErrorDo:` with the new semantics.
 - deprecate `ifError:` to redirect people to the new one

Then eventually we could remove the old one.
What do you think?
That would at least remove part of the "fear component" that incompatibility brings :)

**Comment by guillep** [comment]
News on this front? I think the deprecation is important (and easy!) for this case...

**Comment by guillep** [comment]
I see no activity here.
I'll close it. Please reopen if necessary.

**Comment by Bajger** [comment]
Hi Guille, I'll try to dispatch wrong argument class to new #onErrorDo: method (one would will be on Object with deprecation message, other implementor would be on Error class). We could still preserve original #ifError: method on BlockClosure. I'll add adequate tests. (wrong argument count and wrong argument type). How can I reopen issue again? Only someone from core devs can do that?

**Comment by guillep** [comment]
Thanks for the follow up! I'll reopen it.

> We could still preserve original #ifError: method on BlockClosure

I'm ok with deprecating `ifError:` too ^^. My point is that many external libraries may use this method, and we are just breaking them when it when we could make it much more nicely...

**Comment by guillep** [comment]
@Bajger I see the PR was issued by another user. Give it a try, but maybe it's just easier to do a new PR, as you feel it better.

**Comment by Bajger** [comment]
Can I just clone Kamil's repository with his commit in PR? I guess it's doable, right? But I would need to commit to his repo & PR...

**Comment by kamk** [comment]
Uff, I screwed up something when trying to merge upstream of current Pharo 9.0 so I hopefully have reversed it. Now I have ready `ifErrorDo:` logic so should I commit and push it to my repo?

**Comment by guillep** [comment]
You should be able to do it with some easy git / iceberg manipulations.
For example, if you go up and click on the PRd branch you can get the repository and branch that were issued.

image

Then you may:
 - add that repository url as remote
 - checkout the branch
 - push it to your repository

If you want to do it with iceberg, you need to do some additional manipulations (that are necessary just for contributing pharo, and not for other projects):
 - add that repository url as remote
 - checkout the branch, but selecting in the checkout preview to "Not load any packages" (there is a drop down)
 - merge latest Pharo90 branch into the branch (proceed in the warning)
 - then you will have a branch that has these changes + merged with latest pharo changes.

**Comment by Bajger** [comment]
Hi Kamil, proposed solution from Guille of using new #ifErrorDo: and deprecating #ifError: seems to be more convenient, so I agree. Tests just needs to be updated accordingly (even using testing raising deprecation, if ifError: is sent). You can do new PR, if you have issues with continuing on existing PR.

**Comment by guillep** [comment]
Wow thank you both for the discussion and the effort! I'm going to pass on the PR changes fast and let's wait for the CI :)

**Comment by Bajger** [comment]
Thank you Kamil!

File: src/Kernel/BlockClosure.class.st
Comment (Bajger) [inline src/Kernel/BlockClosure.class.st:358]: Eventually, there could be executable examples created from this comment by using #>>> method.

File: src/Morphic-Base/Form.extension.st
Comment (Bajger) [inline src/Morphic-Base/Form.extension.st:21]: Are these changes relevant to #ifError: refactoring? Or it comes from different commit...

**Review by Bajger — APPROVED** [review]
There are some changes that probably not relevant, but they might be from different commit. But it is mergeable, so problem shouldn't occur, right?

File: src/Morphic-Base/Form.extension.st
Comment (kamk) [reply src/Morphic-Base/Form.extension.st:21]: No, it's mystery to me how did this change get there.

Comment (guillep) [reply src/Morphic-Base/Form.extension.st:21]: It's just the order that got changed. Probably just a change in the tonal format. I think its no biggy :)

**Review by guillep — APPROVED** [review]
Good! Thanks both!

**Comment by astares** [comment]
Why "if" instead of #onErrorDo: ? It dispatches to #on:do: so we should keep the "on"

**Comment by guillep** [comment]
I think this is just because historically the deprecated method was called `ifError:`. I'm not against renaming it as `onErrorDo:`

**Comment by kamk** [comment]
It could also be defined as call to `ifErrorDo:` if this is kind of redudancy is okay.

**Pull request merged**
